test/fixtures/es-module-loaders/register-programmatically-loader-resolve.mjs

```diff
@@ -1,4 +1,3 @@
-import * as fixtures from '../../common/fixtures.mjs';
 import { register } from 'node:module';
 
-register(fixtures.fileURL('es-module-loaders', 'loader-resolve-passthru.mjs').href);
+register('./test/fixtures/es-module-loaders/loader-resolve-passthru.mjs');
```

Comment:
How I could get this working? If I commit your suggestion the tests will fail, the loader path will become something like this:

```shell
/Users/user/home/node/loader-resolve-passthru.mjs
```

Reply:
It needs to resolve the URL from the URL of the caller, not the cwd. Something's wrong with the parentURL I think

Reply:
A while ago I made this change using `dirname()` function to create the `parentURL` directly in the `register()` function and pass it as a parameter along with the `specifier`:

https://github.com/nodejs/node/pull/46826#discussion_r1137978103

Maybe we should discuss it a little more @GeoffreyBooth?

In my opinion, it feels more natural to resolve the `parentURL` as `import()` does.

Reply:
I haven't given it a lot of thought, but what about supporting a callback?

```
register(() => import('../loaders/loader.js'));
```

Reply:
I think we should stick to `import()`'s behaviour (and that would probably be the easiest to do, so win-win).

For a callback… 🤔 that seems over-complicated. If it comes to it, sure.

Reply:
We can’t pass callbacks cross thread, so that’s not an option.

Reply:
Right, then `register(new URL('../loaders/loader.js', import.meta.url))`

Reply:
So absolute URLs only, no support for `node_modules` resolution?

Reply:
Why can't CJS users just leverage `__dirname`?

Reply:
I'm talking about bare specifiers, e.g. `register('ts-node')`, that's what I mean by `node_modules` resolution.

Reply:
When would it be the case that parentUEL isn't available? I can see in the implementation that it's accounted for and cwd is substituted, but can that actually ever occur?

Reply:
? I think we're talking about two completely different things. The problem with using cwd is described in https://github.com/nodejs/node/pull/46826#discussion_r1224164731 and in https://github.com/nodejs/node/pull/46826#discussion_r1224164866.

Reply:
I'm thinking we should be able to leverage however `import.meta.resolve` works. I tried following that setup, and holy crap what a meandering trek. At some point upstream, it receives data from V8 (thanks Antoine for the insight!), which I think happens via `internal_binding()`; I think we should be able to piggyback off that for `register`.

Reply:
If that turns out to not actually be possible, then couldn't we just say they need to provide a resolved url? I think if the user passed the output of `import.meta.resolve()` to `register()`, that would do it.

<details>
<summary>specifier tests</summary>

```console
mkdir test-import-meta-resolve && cd $_
npm i ts-node
node \
  --experimental-import-meta-resolve \
  --input-type=module \
  --eval 'console.log(await import.meta.resolve("ts-node"))'

# file:///[…]/test-import-meta-resolve/node_modules/ts-node/dist/index.js

touch foo.js
mkdir nested && cd $_
node \
  --experimental-import-meta-resolve \
  --input-type=module \
  --eval 'console.log(await import.meta.resolve("../foo.js"))'

# file:///[…]/test-import-meta-resolve/foo.js

touch bar.js
node \
  --experimental-import-meta-resolve \
  --input-type=module \
  --eval 'console.log(await import.meta.resolve("./bar.js"))'

# file:///[…]/test-import-meta-resolve/nested/bar.js
```
</details>

This doesn't look _too_ cumbersome:

```js
import { register } from 'node:module';

register(await import.meta.resolve('ts-node'));

await import('./foo.ts');
```

Note the key difference between `import.meta.resolve()` and `new URL(specifier, import.meta.url)` being `import.meta.resolve()` resolves bare module specifiers (eg `ts-node` works via import.meta.resolve but not new URL).

And that's how it works for [worker threads](https://nodejs.org/api/worker_threads.html#new-workerfilename-options) (which, admittedly, is very, very annoying DX, but at least it's precedent).

Reply:
As I was saying in https://github.com/nodejs/node/pull/46826#discussion_r1224403769, that leaves CJS users behind, which is not great IMO (but not a blocker for this PR)

Reply:
Agh, right, sorry. Too many threads for the same thing.

Why does that leave CJS users behind? With the same setup as above

```console
node -e 'const { pathToFileURL } = require("node:url"); console.log(pathToFileURL(require.resolve("ts-node")).href)'

# file:///[…]/test-import-meta-resolve/node_modules/ts-node/dist/index.js

cd nested

node -e 'const { pathToFileURL } = require("node:url"); console.log(pathToFileURL(require.resolve("../foo.js")).href)'

# file:///[…]/test-import-meta-resolve/foo.js

node -e 'const { pathToFileURL } = require("node:url"); console.log(pathToFileURL(require.resolve("./bar.js")).href)'

# file:///[…]/test-import-meta-resolve/nested/bar.js
```

Reply:
`require.resolve` would resolve the `"require"` condition, when we want the `"import"` condition.

Reply:
Does that matter? A loader can be CJS.

They could potentially fall afoul of the dual package hazard, but I think that's out of scope for this.

Reply:
I think it does matter, otherwise why have a exports map if Node.js doesn't comply with it 🤷‍♂️  maybe we could introduce a new `"register"` condition btw, so users would not need to specify `ts-node/register`.

I agree it's not on scope for this PR.

Reply:
Possibly beating a dead horse (it seems we've reached consensus?): it seems to me node is complying with it. The module is being loaded from a cjs file, so it gets the cjs export. What's incorrect there?

Reply:
It's loaded from the loader thread, not from the module calling `register()`.